cni/pkg/taint/taint.go

```diff
@@ -0,0 +1,387 @@
+// find all pods with selectors and namespaces defined in config map
+func (ts TaintSetter) ListCandidatePods() (list v1.PodList, err error) {
+	list.Items = []v1.Pod{}
+	for _, config := range ts.configs {
+		var rawList *v1.PodList
+		selectors := strings.Split(config.LabelSelectors, ",")
+		for _, selector := range selectors {
+			rawList, err = ts.Client.CoreV1().Pods(config.namespace).List(context.TODO(), metav1.ListOptions{
+				LabelSelector: selector,
+			})
+			if err != nil {
+				return
+			}
+			for _, pod := range rawList.Items {
+				if ts.validTolerance(pod) {
+					list.Items = append(list.Items, pod)
+				}
+			}
+		}
+	}
+	return
+}
+func (ts TaintSetter) HasReadinessTaint(node *v1.Node) bool {
+	taintExists := false
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == TaintName && taint.Effect == v1.TaintEffectNoSchedule {
+			taintExists = true
+			break
+		}
+	}
+	return taintExists
+}
+
+//assumption: order of taint is not important
+func (ts TaintSetter) RemoveReadinessTaint(node *v1.Node) error {
+	updatedTaint := deleteTaint(node.Spec.Taints, &v1.Taint{Key: TaintName, Effect: v1.TaintEffectNoSchedule})
+	node.Spec.Taints = updatedTaint
+	updatedNodeWithTaint, err := ts.Client.CoreV1().Nodes().Update(context.TODO(), node, metav1.UpdateOptions{})
+	if err != nil || updatedNodeWithTaint == nil {
+		return fmt.Errorf("failed to update node %v after adding taint: %v", node.Name, err)
+	}
+	log.Infof("Successfully removed taint on node %v", updatedNodeWithTaint.Name)
+	return nil
+}
+
+// taint node with specific taint name with effect of no schedule
+//do nothing if it already have the readiness taint
+func (ts TaintSetter) AddReadinessTaint(node *v1.Node) error {
+	var taintsExists bool
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == TaintName {
+			taintsExists = true
+			break
+		}
+	}
+	if taintsExists {
+		log.Debugf("%v already present on node %v", TaintName, node.Name)
+		return nil
+	}
+	node.Spec.Taints = append(node.Spec.Taints, v1.Taint{
+		Key:    TaintName,
+		Effect: v1.TaintEffectNoSchedule,
+	})
+	updatedNodeWithTaint, err := ts.Client.CoreV1().Nodes().Update(context.TODO(), node, metav1.UpdateOptions{})
+	if err != nil || updatedNodeWithTaint == nil {
+		return fmt.Errorf("failed to update node %v after adding taint: %v", node.Name, err)
+	}
+	log.Infof("Successfully added taint on node %v", updatedNodeWithTaint.Name)
+	return nil
+}
+
+//get all critical pods in a node
+func (ts TaintSetter) GetAllCriticalPodsInNode(node *v1.Node) (podlist v1.PodList, err error) {
+	candidates, err := ts.ListCandidatePods()
+	if err != nil {
+		return podlist, err
+	}
+	podlist.Items = []v1.Pod{}
+	for _, pod := range candidates.Items {
+		if pod.Spec.NodeName == node.Name {
+			podlist.Items = append(podlist.Items, pod)
+		}
+	}
+	return podlist, nil
+}
+
+//return true if labels contain at least one selector in labelselector class
+func containsLabel(labels map[string]string, selector string) bool {
```

Comment:
Should use k8s libraries like `ParseToLabelSelector`, do not invent it ourselves